lib/eclipse/include/opm/parser/eclipse/EclipseState/AquiferCT.hpp

```diff
@@ -0,0 +1,155 @@
+#define OPM_AQUIFERCT_HPP
+
+/*
+  The AquiferCT which stands for AquiferCarterTracy is a data container object meant to hold the data for the aquifer carter tracy model.
+  This includes the logic for parsing as well as the associated tables. It is meant to be used by opm-grid and opm-simulators in order to
+  implement the Carter Tracy analytical aquifer model in OPM Flow.
+*/
+
+#include <opm/parser/eclipse/EclipseState/EclipseState.hpp>
+#include <opm/parser/eclipse/Parser/ParserKeywords/A.hpp>
+#include <opm/parser/eclipse/Deck/Deck.hpp>
+#include <opm/parser/eclipse/Deck/DeckItem.hpp>
+#include <opm/parser/eclipse/Deck/DeckRecord.hpp>
+#include <opm/parser/eclipse/Deck/DeckKeyword.hpp>
+
+#include <opm/parser/eclipse/EclipseState/Tables/Aqudims.hpp>
+#include <opm/parser/eclipse/EclipseState/Tables/TableContainer.hpp>
+#include <opm/parser/eclipse/EclipseState/Tables/AqutabTable.hpp>
+#include <boost/concept_check.hpp>
+
+namespace Opm {
+
+    class AquiferCT {
+        public:
+
+            struct AQUCT_data{
+
+                    // Aquifer ID
+                    int aquiferID;
+                    // Table IDs
+                    int inftableID, pvttableID;
+                    // Perforation cell id
+                    std::vector<int> cell_id;
+                    // Variables constants
+                    double  phi_aq , //aquifer porosity
+                            d0,   //aquifer datum depth
+                            C_t , //total compressibility
+                            r_o , //aquifer inner radius
+                            k_a , //aquifer permeability
+                            c1, // 0.008527 (METRIC, PVT-M); 0.006328 (FIELD); 3.6 (LAB)
+                            h , //aquifer thickness
+                            theta , //angle subtended by the aquifer boundary
+                            c2 ; //6.283 (METRIC, PVT-M); 1.1191 (FIELD); 6.283 (LAB).
+                    std::vector<double> td, pi;
+            };
+
+            AquiferCT(const EclipseState& eclState, const Deck& deck);
+
+            std::vector<AquiferCT::AQUCT_data> getAquifers();
+
+            const double getAqPorosity(size_t aquiferIndex);
+      
+    private:
+        
+        std::vector<AquiferCT::AQUCT_data> m_aquifers;
+
+        // Initialize function
+        inline std::vector<AquiferCT::AQUCT_data>
+        init_object(const EclipseState& eclState, const Deck& deck)
+        {
+            if (!deck.hasKeyword("AQUCT")){
+                std::cout << "Nothing is shown! Where is AQUCT!????" << std::endl;
+            }
+
+            const auto& aquctKeyword = deck.getKeyword("AQUCT");
+
+            std::vector<AquiferCT::AQUCT_data> aquctParams;
+            // Resize the parameter vector container based on row entries in aquct
+            // We do the same for aquifers too because number of aquifers is assumed to be for each entry in aquct
+            aquctParams.resize(aquctKeyword.size());
```

Comment:
No - avoid resizing. Instead the code should be rewritten to create `AquiferCT::AQUCT_data` instances in the `for` loop below, and those should be added to the `aquctParams` vector using `push_back`or `emplace_back`.

Reply:
Why avoid the resize in this situation if I may ask? Because if I already know the number of elements the vector container should have prior to the for loop, it is still safe as I am also using the at function, no? I am genuinely curious as I am not a super C++ programmer myself, and would like to learn more.

Reply:
What you have done is certainly safe; and the best from a performance point view. My argument is not really about C++, but just a general coding principle - where I think your approach is less readable and more difficult to maintain (not very difficult to read/maintain - mind you; but still). An objective argument for side of the story is that you create *couplings* between different parts of the code:
```C++
vector.resize( );         <----------\
...                                  |
...                                  |
for ( ) {                            |
     vector.at( ) = ...;   <---------/
}
```

non-local couplings generally reduce readability and maintainability. But nothing to get religious about.